src/Mod/Sketcher/Gui/DrawSketchHandlerScale.h

```diff
@@ -340,48 +363,68 @@ class DrawSketchHandlerScale: public DrawSketchHandlerScaleBase
+                    && secondIndex != GeoEnum::GeoUndef && thirdIndex != GeoEnum::GeoUndef) {
+                    newConstr->Second = offsetGeoID(secondIndex, firstCurveCreated);
+                    newConstr->Third = offsetGeoID(thirdIndex, firstCurveCreated);
                 }
                 else if ((cstr->Type == Coincident || cstr->Type == Tangent
                           || cstr->Type == Symmetric || cstr->Type == Perpendicular
                           || cstr->Type == Parallel || cstr->Type == Equal || cstr->Type == Angle
                           || cstr->Type == PointOnObject || cstr->Type == InternalAlignment)
-                         && firstIndex >= 0 && secondIndex >= 0
-                         && thirdIndex == GeoEnum::GeoUndef) {
-                    newConstr->Second = firstCurveCreated + secondIndex;
+                         && secondIndex != GeoEnum::GeoUndef && thirdIndex == GeoEnum::GeoUndef) {
+                    newConstr->Second = offsetGeoID(secondIndex, firstCurveCreated);
                 }
-                else if ((cstr->Type == Radius || cstr->Type == Diameter) && firstIndex >= 0) {
+                else if (cstr->Type == Radius || cstr->Type == Diameter) {
                     newConstr->setValue(newConstr->getValue() * scaleFactor);
                 }
                 else if ((cstr->Type == Distance || cstr->Type == DistanceX
                           || cstr->Type == DistanceY)
-                         && firstIndex >= 0 && secondIndex >= 0) {
-                    newConstr->Second = firstCurveCreated + secondIndex;
+                         && secondIndex != GeoEnum::GeoUndef) {
+                    newConstr->Second = offsetGeoID(secondIndex, firstCurveCreated);
                     newConstr->setValue(newConstr->getValue() * scaleFactor);
                 }
-                else if ((cstr->Type == Block || cstr->Type == Weight) && firstIndex >= 0) {
-                    newConstr->First = firstCurveCreated + firstIndex;
-                }
-                else {
-                    continue;
-                }
+                // (cstr->Type == Block || cstr->Type == Weight)
 
                 ShapeConstraints.push_back(std::move(newConstr));
             }
         }
     }
+    bool skipConstraint(const Constraint* constr)
+    {
+        return (constr->First == GeoEnum::GeoUndef)
+            || (!allowOriginConstraint
+                && (constr->First == GeoEnum::VAxis || constr->First == GeoEnum::HAxis
+                    || constr->Second == GeoEnum::VAxis || constr->Second == GeoEnum::HAxis
+                    || constr->Third == GeoEnum::VAxis || constr->Third == GeoEnum::HAxis))
+            || (constr->First != GeoEnum::GeoUndef && constr->First <= GeoEnum::RefExt)
+            || (constr->Second != GeoEnum::GeoUndef && constr->Second <= GeoEnum::RefExt)
+            || (constr->Third != GeoEnum::GeoUndef && constr->Third <= GeoEnum::RefExt);

```

Comment:
That is a biiig condition! I don't tihnk that it is worth  reworking, but it could use at least a comment above explaining what it aims to achieve.